### PR TITLE
Random fixes

### PR DESCRIPTION
**Pull request by franz**
1) fix some random annoying warnings from CMake,

2) remove a redundant (for HSA at least) mkdir in clEnqueueNDRangeKernel() - improves performance of small kernel enqueues

### CODE REVIEW

File: lib/CL/clEnqueueNDRangeKernel.c [hunk 1]
```diff
@@ -237,20 +237,18 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
     CL_INVALID_EVENT_WAIT_LIST);
 
   char cachedir[POCL_FILENAME_LENGTH];
-  pocl_cache_make_kernel_cachedir_path (cachedir, kernel->program,
-                                        realdev, kernel,
-                                        local_x, local_y, local_z);
-
   int realdev_i = pocl_cl_device_to_index(kernel->program, realdev);
+  assert(realdev_i >= 0);
+  pocl_cache_kernel_cachedir_path(cachedir, kernel->program, realdev_i, kernel, "", local_x, local_y, local_z);
```
Comment (pjaaskel) [inline lib/CL/clEnqueueNDRangeKernel.c]: Please mind the style. Space before ( and probably too long line?

File: lib/CL/clEnqueueNDRangeKernel.c [hunk 2]
```diff
@@ -237,20 +237,18 @@ POname(clEnqueueNDRangeKernel)(cl_command_queue command_queue,
...
+
   if (kernel->program->source || kernel->program->binaries[realdev_i])
     {
 #ifdef OCS_AVAILABLE
       // SPMD devices already have compiled at this point
       if (realdev->spmd)
         error = CL_SUCCESS;
       else
-        error = pocl_llvm_generate_workgroup_function(realdev,
```
Comment (pjaaskel) [inline lib/CL/clEnqueueNDRangeKernel.c]: Ditto.

**Comment by franz** [comment]
OK, fixed style

File: lib/CL/clBuildProgram.c [hunk 3]
```diff
@@ -127,11 +127,11 @@ program_compile_dynamic_wg_binaries(cl_program program)
 
       for (i=0; i < program->num_kernels; i++)
         {
-          pocl_cache_make_kernel_cachedir_path(cachedir, program, device,
-                                               program->default_kernels[i],
-                                               0,0,0);
+          pocl_cache_kernel_cachedir_path(cachedir, program, device_i,
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Check parenthesis in the call. Should be func ();

File: lib/CL/clBuildProgram.c [hunk 4]
```diff
@@ -127,11 +127,11 @@ program_compile_dynamic_wg_binaries(cl_program program)
...
+                                          program->default_kernels[i],
+                                          "", 0,0,0);
 
-          errcode = pocl_llvm_generate_workgroup_function(device,
+          errcode = pocl_llvm_generate_workgroup_function(cachedir, device,
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Ditto.

File: lib/CL/pocl_llvm_api.cc [hunk 5]
```diff
@@ -1385,7 +1385,7 @@ kernel_library
 /* This is used to control the kernel we want to process in the kernel compilation. */
 extern cl::opt<std::string> KernelName;
 
-int pocl_llvm_generate_workgroup_function(cl_device_id device, cl_kernel kernel,
+int pocl_llvm_generate_workgroup_function(char* kernel_cachedir, cl_device_id device, cl_kernel kernel,
```
Comment (pjaaskel) [inline lib/CL/pocl_llvm_api.cc]: Long line.

**Comment by pjaaskel** [comment]
Some style issues found still.

File: CMakeLists.txt [hunk 6]
```diff
@@ -771,7 +758,19 @@ if(ENABLE_TCE)
 
   set(TCE_DEVICE_CL_VERSION "120")
 
+  if("${LLVM_CXXFLAGS}" MATCHES "-fno-rtti")
+    message(WARNING "TCE is enabled but your LLVM was not built with RTTI. You should rebuild LLVM with 'make REQUIRES_RTTI=1'. See the INSTALL file for more information.")
```
Comment (pjaaskel) [inline CMakeLists.txt]: Is it possible to split this line in the Cmake scripting language?

File: include/pocl_cache.h [hunk 7]
```diff
@@ -88,14 +88,14 @@ int pocl_cache_write_descriptor(cl_program   program,
...
 
-int pocl_cache_make_kernel_cachedir_path(char*        kernel_cachedir_path,
-                                         cl_program   program,
-                                         cl_device_id device,
-                                         cl_kernel    kernel,
-                                         size_t       local_x,
-                                         size_t       local_y,
-                                         size_t       local_z);
-
+void pocl_cache_kernel_cachedir_path (char* kernel_cachedir_path,
```
Comment (pjaaskel) [inline include/pocl_cache.h]: Check C style func ().

File: lib/CL/clBuildProgram.c (hunk 3 above)
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Not fixed.

File: lib/CL/clBuildProgram.c (hunk 4 above)
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Ditto.

File: lib/CL/pocl_llvm_api.cc (hunk 5 above)
Comment (pjaaskel) [inline lib/CL/pocl_llvm_api.cc]: Still.

**Comment by pjaaskel** [comment]
After pulling your branch, I'm unable to build pocl:

```
[ 97%] Building LL to LLVM bitcode /home/visit0r/src/pocl-public/test-pr-cmake-llvm-3.9-build/lib/kernel/host/bdver3/svm_atomics_x86_64_llvm3_9.ll.bc
[ 98%] Linking & optimizing Kernel bitcode /home/visit0r/src/pocl-public/test-pr-cmake-llvm-3.9-build/lib/kernel/host/kernel-x86_64-unknown-linux-gnu-bdver3.bc
WARNING: Linking two modules of different data layouts: '/home/visit0r/src/pocl-public/test-pr-cmake-llvm-3.9-build/lib/kernel/host/bdver3/barrier.ll.bc' is '' whereas 'llvm-link' is 'e-m:e-i64:64-f80:128-n8:16:32:64-S128'

[ 98%] Built target kernel_host_bdver3
Scanning dependencies of target kernellib_hash
[ 98%] Generating SHA1 of all kernel libs...
-- Renaming /home/visit0r/src/pocl-public/test-pr-cmake-llvm-3.9-build/kernellib_hash.h.new to /home/visit0r/src/pocl-public/test-pr-cmake-llvm-3.9-build/kernellib_hash.h
[ 98%] Built target kernellib_hash
Scanning dependencies of target pocl_cache
[100%] Building C object lib/CL/CMakeFiles/pocl_cache.dir/pocl_cache.c.o
[100%] Built target pocl_cache
Scanning dependencies of target pocl
[100%] Linking CXX shared library libpocl.so
/usr/bin/ld: CMakeFiles/libpocl_unlinked_objs.dir/clEnqueueNDRangeKernel.c.o: unrecognized relocation (0x2a) in section `.text'
/usr/bin/ld: final link failed: Bad value
```

**Comment by franz** [comment]
@pjaaskel  is that with a clean build tree ? i've never seen that error..

**Comment by pjaaskel** [comment]
Yeah. At least LLVM 3.9 build complains about that even after rm -fr buildtree.

**Comment by pjaaskel** [comment]
Probably something local as it now keeps failing all the time randomly with a similar issue. Even with clean master.

**Comment by pjaaskel** [comment]
Can you please check the conflicts?

**Comment by franz** [comment]
Rebased on top of master

File: lib/CL/clBuildProgram.c [hunk 8]
```diff
@@ -139,12 +139,11 @@ program_compile_dynamic_wg_binaries(cl_program program)
               local_z = kernel->reqd_wg_size[2];
             }
 
-          pocl_cache_make_kernel_cachedir_path (cachedir, program, device,
-                                                kernel, local_x, local_y, local_z);
+          pocl_cache_kernel_cachedir_path(cachedir, program, device_i, kernel,
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Missing space before (

File: lib/CL/clBuildProgram.c [hunk 9]
```diff
@@ -139,12 +139,11 @@ program_compile_dynamic_wg_binaries(cl_program program)
...
+                                          "", local_x, local_y, local_z);
 
-          errcode
-            = pocl_llvm_generate_workgroup_function
-            (device, kernel, local_x, local_y, local_z);
+          errcode = pocl_llvm_generate_workgroup_function(cachedir, device, kernel,
```
Comment (pjaaskel) [inline lib/CL/clBuildProgram.c]: Ditto.

**Comment by franz** [comment]
fixed & rebased

**Pull request merged**